doc/development/06_ca_manager.md

```diff
@@ -0,0 +1,976 @@
+            // 2. this ensures that there are no apparent data issues
+            //
+            // If there are issues, then complain and try to recover.
+            error!(
+                "Could not warm up cache, data seems corrupt. Will try to recover!! Error was: {}",
+                e
+            );
+            ca_store.recover()?;
+        }
+
+        // Create the `CaObjectStore` that is responsible for maintaining CA objects: the `CaObjects`
+        // for a CA gets copies of all ROAs and delegated certificates from the `CertAuth` and is responsible
+        // for manifests and CRL generation.
+        let ca_objects_store = Arc::new(CaObjectsStore::disk(config.clone(), signer.clone())?);
+
+        // Register the `CaObjectsStore` as a pre-save listener to the 'ca_store' so that it can update
+        // its ROAs and delegated certificates and/or generate manifests and CRLs when relevant changes
+        // occur in a `CertAuth`.
+        ca_store.add_pre_save_listener(ca_objects_store.clone());
+
+        // Register the `MessageQueue` as a post-save listener to 'ca_store' so that relevant changes in
+        // a `CertAuth` can trigger follow up actions. Most importantly: synchronize with a parent CA or
+        // the RPKI repository.
+        ca_store.add_post_save_listener(mq);
+
+        // Create the status store which will maintain the last known connection status between each CA
+        // and their parent(s) and repository.
+        let status_store = StatusStore::new(&config.data_dir, STATUS_DIR)?;
+
+        // Create the per-CA lock structure so that we can guarantee safe access to each CA, while allowing
+        // multiple CAs in a single Krill instance to interact: e.g. a child can talk to its parent and they
+        // are locked individually.
+        let locks = Arc::new(CaLocks::default());
+
+        Ok(CaManager {
+            ca_store: Arc::new(ca_store),
+            ca_objects_store,
+            status_store: Arc::new(Mutex::new(status_store)),
+            locks,
+            config,
+            signer,
+        })
+    }
+}
+```
+
+To illustrate how the event listening is used here: If a `CaEvtDet::RoasUpdated` event occurs,
+then this will trigger:
+1. (pre-save) that the `CaObjectsStore` updates the ROAs held by the CA,
+   and generates a new CRL and manifest; and
+2. (post-save) that the `MessageQueue` notices, and schedules a task to
+   synchronize the contents of the `CaObjects` with the CA's repository.
+
+
+Asynchronous Actions
+--------------------
+
+The aforementioned `MessageQueue` is shared with the `Scheduler` which is owned
+by `KrillServer`. The `Scheduler` uses the `clokwerk.rs` library to schedule several
+background jobs in Krill. One of these jobs watches the `MessageQueue` for queued
+tasks, added here because of events that occurred.
+
+This background job has access to its own `Arc<CaManager>` and `Arc<RepositoryManager>`,
```

Comment:
For the less Rust aware this may be misinterpreted as saying that the background job has its own separate copies of the CaManager and RepositoryManager, while actually it has references to the one and only instances of these managers. Perhaps this isn't worth adjusting as you should know Rust to work on Krill, but if you're an interested 3rd party trying to get an idea of how Krill works but are not fluent in Rust does this perhaps merit some clarification?

Reply:
I think I'm one of those interested parties. I know that I still have to read about std::sync::Arc to really understand that is happening here.

Reply:
pushed:
```
A Krill instance only has a single (singleton) `CaManager` and `RepositoryManager`, which
are kept as `Arc<CaManager>` and `Arc<RepositoryManager>` so that they (well the reference)
can easily be shared and copied.
```